README.md
# animation_director

A package to create nice and smooth animations for flutter

## Introduction

A simple package to build beautiful and smooth animations for flutter framework. By using this package, you don't need 
to bother yourself with `AnimationController`s, `AnimationTweens` and also timing and of those animations.

Here there are some simple examples of what I've made with this package in 5 minutes for each one.


![](showcase/animator-d1-h.gif)
![](showcase/animator-m3x3-h.gif)
![](showcase/hamburger-h.gif)
![](showcase/mr-fastfood-h.gif)
![](showcase/taxi-h.gif)


The following descriptions may sounds complicated but it's pretty simple, sorry for my bad english, I think if you jump to 
example directory and take a look at the codes, you'll find out it's really easy to use



## ActorWidget

Each object on the screen is called an `ActorWidget`.

`ActorWidget` has two required properties:

```dart
    @required String name;
    @required List<ActorAction> actions;
```

each `ActorAction` is an individual animation of the `ActorWidget`.

All `ActorWidget`s in your page run simultaneously.
All `ActorAction`s inside of an `ActorWidget` run sequentially, but you can control how much it waits before starts the animation
using `waitBeforeStart`


```dart
  // ___________________________________
  ActorWidget(                          |
    name: 'Actor1',                     |
    actions: [                          |
      // ___________                    |
      ActorAction(  |                   |
        ...         |                   |
      ),            |                   |
      ActorAction(  |                   |
        ...         | (Sequentially)    | (Simultaneously)
      ),            |                   |
      ActorAction(  |                   |
        ...         |                   |
      )             |                   |
      // ___________|                   |
  ]),                                   |
  ActorWidget(),                        | 
  ActorWidget(),                        |
  ActorWidget(),                        |
  // ___________________________________|
  
```



# ActorAction Features

Each `ActorAction` has following features

```dart
    List<String> group; // you can define one or multi group for action.
    Function onCompleted;
    Function onStart;
    Duration waitBeforeStart; // the amount of time the current action must wait to start, after the previous action is finished
    ActorPath path;
    ActorPosition position;
    ActorOpacity opacity;
    ActorCharacter character;
    ActorRotation rotation;
    ActorScale scale;
```

if you want to animate some of your `ActorWidget`s, you can give a name to the actions of your actor. then start your `AnimationDirector` by that group.
Using this, only `ActorWidget`s which has the given `group` animate.  
This is useful when you want to trigger an animation manually. see Example Menu3x3.


## ActorCharacter

The actor character! your actor is represented by `ActorCharacter`.

`ActorCharacter` has the following properties:

```dart
    Duration duration;
    Curve curve;
    double width;
    double height;
    Widget child;
    String clipPath; // the string of the clip path
    EdgeInsets padding;
    EdgeInsets margin;
    BoxDecoration decoration;
    BoxDecoration foregroundDecoration;
```

**Note**: `clipPath` type is string! if follows svg codes rule. and for now only supports `M`, `L`, `A`, `Q`.

In order to generate a path you can use [FlutterDev.site Path Creator](http://flutterdev.site/path-creator). 
Draw your path and copy `The Result` to the `clipPath` property



## ActorPosition

You can animate the position of `Actor` using `ActorPosition`.

`ActorPosition` has the following properties

```dart
    Duration duration;
    Curve curve;
    double top;
    double left;
    double right;
    double bottom;
```


## ActorOpacity

To change the opacity of your `Actor`.

`ActorOpacity` has the following properties

```dart
    Duration duration;
    Curve curve;
    double opacity;
```

## ActorScale

To change the scale of your `Actor`.

`ActorScale` has the following properties

```dart
    Duration duration;
    Curve curve;
    double startScale;
    double finishScale;
```

## ActorRotation

To rotate, flip horizontal or vertical your `Actor` widget, you can use `ActorRotation`.

`ActorRotation` has the following properties

```dart
    Duration duration;
    Curve curve;
    double rotationTurns;
    Alignment alignment;
    bool clockwise;
    double startTween;
    double verticalFlipTurns;
    double horizontalFlipTurns;
```
    

## ActorPath

You can draw a path by using `ActorPath`. and if you define an `ActorCharacter` along with `ActorPath`, the character will
move along the created path.

To draw your path again use [FlutterDev.site Path Creator](http://flutterdev.site/path-creator).

`ActorPath` has the following properties
    
```dart
    Duration duration;
    Curve curve;
    String path; // the String code of the path
    int traversePercentage; // the percentage of `ActorCharacter` movement along the path
    int startPositionInPercent; // the start position of animation in percent
    bool displayPath; // displays the path on the screen, 
    bool displayProgress; // displays the progress of path filling on the screen
    Offset offset; // you can align the position of ActorCharacter widget on the path
    Paint pathStyle; // style of the path line
    Paint progressStyle; // style of the progress line
    int progressLength; // the length of progress line, the default is as equal as the path
    int progressAnimationRepeat; // the repeat times of progress animation
    bool fadingProgress; // to make the tail of the progress line faded
```

# Path Creator

In order to create paths easily head to [FlutterDev.site Path Creator](http://flutterdev.site/path-creator).  
Draw your path and copy the generated code from `The Result`.

## Adaptive Size

By default, the generated path is based on fixed numbers of your path and has the following effect

![](showcase/fixed.png)
 
If you want to draw your path to be adaptive to the screen of any device, you need to select `Adaptive Width`
or `Adaptive Height` or both if you want your path to be fully adaptive.
When you select for example `Adaptive Width`, the drawing-area width will be considered as mobile,tablet,... width so you get following
effect with `Adaptive Width`. (500 for example)

![](showcase/adaptive.png)

But keep in mind, in both cases, the result may be different on different screen-sizes. 



# Example

Let's say we want to animate an orange circle from top-center of the screen to the center. Then after 2 seconds,
move the circle to the bottom of the page.  


The first step is to create our character and position it to top of the page.
```dart
  ActorAction(
    position: ActorPosition(top: 0, left: 0, right: 0),
    character: ActorCharacter(
      child: Center(
        child: Container(
          width: 50,
          height: 50,
          decoration: BoxDecoration(
            color: Colors.orangeAccent,
            borderRadius: BorderRadius.circular(50),
          ),
        ),
      )
    ),
  ),
```

The second step is to move it to the center of the page, so let's add another `ActorAction`
```dart

    ...
    ActorAction(
        position: ActorPosition(
          curve: Curves.elasticOut,
          duration: Duration(milliseconds: 800),
          top: MediaQuery.of(context).size.height / 2 - 25),
    ),
    ...
    
```

And the final step is to move our circle to the bottom of the page after 2 seconds

```dart

    ...
    ActorAction(
        waitBeforeStart: Duration(seconds: 2),
        position: ActorPosition(
          duration: Duration(milliseconds: 800),
          curve: Curves.elasticOut,
          top: MediaQuery.of(context).size.height - 100),
    ),
    ...
    
```

and you get the following result

![](showcase/example-h.gif)



CHANGELOG.md
## [0.0.1] - 03/15/2020

* package uploaded

## [0.1.0] - 03/16/2020

* some necessary updates

## [0.1.1] - 03/16/2020

* some optional updates !
* added 1 more example

example/Taxi.dart
import 'package:flutter/material.dart';
import 'package:animation_director/animation_director.dart';

class TaxiAnimation extends StatefulWidget {
  @override
  _TaxiAnimationState createState() => _TaxiAnimationState();
}

class _TaxiAnimationState extends State<TaxiAnimation> {

  @override
  Widget build(BuildContext context) {
    return AnimationDirector(
      repeatTimes: 100,
      speed: 1,
      autoStart: true,
      onCompleted: (int repeated, Duration runtime) {
        print("FINISHED - REPEATED $repeated in ${runtime.inMilliseconds}");
      },
      cast: [
        ActorWidget(
            name: 'Taxi',
            actions: [
              ActorAction(
                position: ActorPosition(top: 85, left: 10),
                  character: ActorCharacter(
                    width: 170,
                    child: Image.asset('assets/images/taxi/taxi.png')
                  )
              ),
              ActorAction(
                  position: ActorPosition(top: 84.7),

              ),
              ActorAction(
                  position: ActorPosition(top: 85.3),
              )
            ]),

      ],
    );
  }
}




class Taxi extends StatefulWidget {
  @override
  _TaxiState createState() => _TaxiState();
}

class _TaxiState extends State<Taxi> {

  @override
  Widget build(BuildContext context) {

    return Stack(
      children: <Widget>[
        AnimationDirector(
          repeatTimes: 1,
          speed: 1,
          autoStart: true,
          onCompleted: (int repeated, Duration runtime) {
            print("FINISHED - REPEATED $repeated in ${runtime.inMilliseconds}");
          },
          cast: [

            ActorWidget(
                name: 'Taxi',
                actions: [
                  ActorAction(
                      position: ActorPosition(top: 20, left: 0, right: 0),
                      opacity: ActorOpacity(opacity: 0),
                      character: ActorCharacter(
                          child: Center(child: Text('Taxi', style: TextStyle(fontSize: 75),))
                      )
                  ),
                  ActorAction(
                      waitBeforeStart: Duration(milliseconds: 1000),
                      opacity: ActorOpacity(opacity: 1),
                      position: ActorPosition(top: 190, duration: Duration(milliseconds: 800))
                  ),
                ]
            ),
            ActorWidget(
                name: 'Numbers',
                actions: [
                  ActorAction(
                      position: ActorPosition(top: 280, left: 0, right: 0),
                      opacity: ActorOpacity(opacity: 0),
                      character: ActorCharacter(
                          child: Center(child: Column(
                            children: <Widget>[
                              Text('+521 (xxx)-xxx-xx-xx', style: TextStyle(fontSize: 18),),
                              Text('+521 (xxx)-xxx-xx-xx', style: TextStyle(fontSize: 18),),
                            ],
                          ))
                      )
                  ),
                  ActorAction(
                      waitBeforeStart: Duration(milliseconds: 1500),
                      opacity: ActorOpacity(opacity: 1, duration: Duration(milliseconds: 800)),
                  ),
                ]
            ),
            ActorWidget(
                name: 'Sky',
                actions: [
                  ActorAction(
                      position: ActorPosition(top: 0, right: 0, left: 0),
                      character: ActorCharacter(
                          height: 180,
                          decoration: BoxDecoration(
                              color: Color(0xff1b1e23),
                          )
                      )
                  ),
                ]
            ),
            ActorWidget(
                name: 'Sun',
                actions: [
                  ActorAction(
                      position: ActorPosition(top: 130, right: 80),
                      opacity: ActorOpacity(opacity: 0),
                      character: ActorCharacter(
                          height: 30,
                          width: 30,
                          decoration: BoxDecoration(
                              color: Color(0xfff8f8f8),
                              borderRadius: BorderRadius.circular(50),
                              boxShadow: [
                                BoxShadow(color: Colors.white, spreadRadius: 3, blurRadius: 5)
                              ]
                          )
                      )
                  ),
                  ActorAction(
                    waitBeforeStart: Duration(milliseconds: 1200),
                    opacity: ActorOpacity(opacity: 1),
                    position: ActorPosition(top: 20,
                      curve: Curves.easeOutBack,
                      duration: Duration(milliseconds: 1000),),
                  ),
                ]
            ),
            ActorWidget(
              name: 'Road',
              actions: [
                ActorAction(
                    position: ActorPosition(top: 170, left: 10, right: 10),
                    character: ActorCharacter(
                        child: Image.asset('assets/images/taxi/road.png')
                    )
                ),
              ]
            ),
            fallingBuilding(
                name: 'Building1',
                image: 'building1',
                wait: 100,
                left: 80
            ),
            fallingBuilding(
                name: 'Building2',
              image: 'building2',
              wait: 400,
              left: 140,
              top: 63,
            ),

            ActorWidget(
                name: 'AnimateTaxi',
                actions: [
                  ActorAction(
                      position: ActorPosition(top: 0, right: -150),
                      character: ActorCharacter(
                          child: Container(
                              width: 100,
                              height: 100,
                              child: TaxiAnimation())
                      )
                  ),
                  ActorAction(
                    position: ActorPosition(
                        duration: Duration(milliseconds: 900),
                        right: 170),
                  )
                ]),
            ActorWidget(
                name: 'Bush1',
                actions: [
                  ActorAction(
                      position: ActorPosition(top: 120, right: 20),
                      scale: ActorScale(startScale: 0, finishScale: 0),
                      character: ActorCharacter(
                          height: 100,
                          child: Image.asset('assets/images/taxi/bush.png')
                      )
                  ),
                  ActorAction(
                    waitBeforeStart: Duration(milliseconds: 200),
                    scale: ActorScale(
                        duration: Duration(milliseconds: 1200),
                        curve: Curves.elasticOut,
                        startScale: 0, finishScale: 0.6, alignment: Alignment.center),
                  ),
                ]
            ),
            ActorWidget(
                name: 'Bush2',
                actions: [
                  ActorAction(
                      position: ActorPosition(top: 120, left: 20),
                      scale: ActorScale(startScale: 0, finishScale: 0),
                      character: ActorCharacter(
                          height: 100,
                          child: Image.asset('assets/images/taxi/bush.png')
                      )
                  ),
                  ActorAction(
                    waitBeforeStart: Duration(milliseconds: 600),
                    scale: ActorScale(
                        duration: Duration(milliseconds: 300),
                        curve: Curves.decelerate,
                        startScale: 0, finishScale: 0.4, alignment: Alignment.center),
                  ),
                ]
            ),
            light(name: 'Light1', wait: 2000),
            light(name: 'Light2', wait: 2600),
            light(name: 'Light1', wait: 5000),
            light(name: 'Light2', wait: 5600),

          ],
        )
      ],
    );
  }


  ActorWidget light({
  String name,
    int wait
}) {
    return ActorWidget(
        name: name,
        actions: [
          ActorAction(
            waitBeforeStart: Duration(milliseconds: wait),
              position: ActorPosition(top: 88, left: 30),
              scale: ActorScale(startScale: 0, finishScale: 0),
              character: ActorCharacter(
                  height: 100,
                  child: Image.asset('assets/images/taxi/light.png')
              )
          ),
          ActorAction(
            scale: ActorScale(
                duration: Duration(milliseconds: 10),
                curve: Curves.decelerate,
                startScale: 0, finishScale: 0.5, alignment: Alignment.center),
          ),
          ActorAction(
            waitBeforeStart: Duration(milliseconds: 300),
            scale: ActorScale(
                duration: Duration(milliseconds: 10),
                curve: Curves.decelerate,
                startScale: 0.5, finishScale: 0, alignment: Alignment.center),
          ),
        ]
    );
  }

  ActorWidget fallingBuilding({
    String name,
    String image,
    int wait = 0,
    double left,
    double top = 48,

}) {
    return ActorWidget(
        name: name,
        actions: [
          ActorAction(
            waitBeforeStart: Duration(milliseconds: wait),
              position: ActorPosition(top: -200, left: left),
              character: ActorCharacter(
                  width: 150,
                  child: Image.asset('assets/images/taxi/${image}.png')
              )
          ),
          ActorAction(
            position: ActorPosition(top: top, duration: Duration(milliseconds: 1000),
                curve: Curves.bounceOut
            ),
          )
        ]);
  }
}


example/Hamburger.dart
import 'package:flutter/material.dart';
import 'package:animation_director/animation_director.dart';

class Hamburger extends StatefulWidget {
  @override
  _HamburgerState createState() => _HamburgerState();
}

class _HamburgerState extends State<Hamburger> {

  Paint progressPaint;

  @override
  void initState() {

    progressPaint = Paint()
      ..strokeWidth = 8.0
      ..color = Colors.white
      ..strokeCap = StrokeCap.round
      ..blendMode = BlendMode.softLight;

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return AnimationDirector(
      repeatTimes: 1,
      speed: 1,
      autoStart: true,
      onCompleted: (int repeated, Duration runtime) {
        print("FINISHED - REPEATED $repeated in ${runtime.inMilliseconds}");
      },
      cast: [
/*          ActorWidget(
                actions: [
                  ActorAction(
                    character: ActorCharacter(
                      width: MediaQuery.of(context).size.width,
                      height: MediaQuery.of(context).size.height,
                      decoration: BoxDecoration(
                        color: Colors.blue.withOpacity(0.7)
                      )
                    )
                  )
                ]
              ),*/
        ActorWidget(
            name: '1',
            actions: [
              ActorAction(
                  position: ActorPosition(top: 100, left: -400),
                  character: ActorCharacter(
                      width: 400,
                      height: 400,
                      decoration: BoxDecoration(
                          color: Colors.orangeAccent,
                          borderRadius: BorderRadius.circular(200)))),
              ActorAction(
                position: ActorPosition(
                    top: 100,
                    left: -230,
                    duration: Duration(milliseconds: 1300),
                    curve: Curves.elasticOut),
              ),
              ActorAction(
                waitBeforeStart: Duration(milliseconds: 1500),
                position: ActorPosition(
                    top: 100,
                    left: -500,
                    duration: Duration(milliseconds: 1300),
                    curve: Curves.elasticOut),
              )
            ]),
        ActorWidget(
            name: '2',
            actions: [
              ActorAction(
                  waitBeforeStart: Duration(milliseconds: 3000),
                  position: ActorPosition(
                      top: 0, left: -MediaQuery.of(context).size.width/2),
                  character: ActorCharacter(
                      width: MediaQuery.of(context).size.width/2,
                      height: MediaQuery.of(context).size.height,
                      decoration: BoxDecoration(color: Colors.deepPurple))),
              ActorAction(
                  position: ActorPosition(
                      left: 0,
                      curve: Curves.easeOutCubic,
                      duration: Duration(milliseconds: 600))),
              ActorAction(
                  character: ActorCharacter(
                      decoration: BoxDecoration(
                          color: Color(0xff077D19)
                      ),
                      curve: Curves.decelerate,
                      duration: Duration(milliseconds: 3000)))
            ]),
        optionCircle(
            traversePercentage: 80,
            pathCurve: Curves.linearToEaseOut,
            pathDuration: Duration(milliseconds: 600),
            waitBefore: Duration(seconds: 1),
            rotateDuration: Duration(milliseconds: 1000),
            rotationTurn: 0,
            image: 'food4'),
        optionCircle(
            traversePercentage: 60,
            pathCurve: Curves.linearToEaseOut,
            pathDuration: Duration(milliseconds: 800),
            waitBefore: Duration(seconds: 1),
            rotateDuration: Duration(milliseconds: 1300),
            rotationTurn: 1,
            image: 'food1'),
        optionCircle(
            traversePercentage: 42,
            pathCurve: Curves.linearToEaseOut,
            pathDuration: Duration(milliseconds: 1000),
            waitBefore: Duration(seconds: 1),
            rotateDuration: Duration(milliseconds: 1200),
            image: 'food3'),
        optionCircle(
            traversePercentage: 20,
            pathCurve: Curves.linearToEaseOut,
            pathDuration: Duration(milliseconds: 1200),
            waitBefore: Duration(seconds: 1),
            rotateDuration: Duration(milliseconds: 1500),
            image: 'food2'),
        itemText(
          curve: Curves.decelerate,
          duration: Duration(milliseconds: 1000),
          top: 130,
          title: 'Hamburger',
          waitBefore: Duration(milliseconds: 3000),
        ),
        itemText(
          curve: Curves.decelerate,
          duration: Duration(milliseconds: 1000),
          top: 230,
          title: 'Fried Egg',
          waitBefore: Duration(milliseconds: 3100),
        ),
        itemText(
          curve: Curves.decelerate,
          duration: Duration(milliseconds: 1000),
          top: 340,
          title: 'Chicken',
          waitBefore: Duration(milliseconds: 3200),
        ),
        itemText(
          curve: Curves.decelerate,
          duration: Duration(milliseconds: 1000),
          top: 440,
          title: 'Sausage',
          waitBefore: Duration(milliseconds: 3300),
        ),
        /*ActorWidget(
                name: '3',
                actions: [
              ActorAction(
                  position: ActorPosition(
                    top: MediaQuery.of(context).size.height,
                    left: 0,
                  ),
                  character: ActorCharacter(
                      width: MediaQuery.of(context).size.width,
                      height: 300,
                      decoration: BoxDecoration(color: Color(0xfffbe9e9)),
                      clipPath:
                      'M 0 400 L 400 400 L 400 50 L 350 30 L 310 50 L 270 10 L 250 40 L 230 30 L 190 30 L 170 20 L 130 20 L 120 50 L 100 60 L 60 40 L 40 20 L 20 30 L 0 20 Z')),
              ActorAction(
                  waitBeforeStart: Duration(seconds: 4),
                  position: ActorPosition(
                      top: MediaQuery.of(context).size.height - 200,
                      duration: Duration(milliseconds: 600),
                      curve: Curves.easeOut))
            ]),*/
/*            ActorWidget(
                name: '4',
                actions: [
              ActorAction(
                  position: ActorPosition(
                    top: MediaQuery.of(context).size.height - 200,
                    left: 0,
                  )),
              ActorAction(
                  waitBeforeStart: Duration(seconds: 5),
                  path: ActorPath(
                      displayProgress: true,
                      progressStyle: progressPaint,
                      progressLength: 400,
                      path:
                      'M 310 50 L 270 10 L 250 40 L 230 30 L 190 30 L 170 20 L 130 20 L 120 50 L 100 60 L 60 40 L 40 20 L 20 30 L 0 20 ',
                      duration: Duration(seconds: 5)))
            ]),
            ActorWidget(
                name: '5',
                actions: [
              ActorAction(
                  position: ActorPosition(
                    top: 20,
                    left: 8,
                  ),
                  character: ActorCharacter(
                      height: 200,
                      width: 300,
                      child: Text(
                        "Burgershop",
                        style: TextStyle(fontSize: 55,
                          color: Colors.greenAccent,
                          shadows: [
                            Shadow(
                              blurRadius: 10.0,
                              color: Colors.white.withOpacity(0.5),
                              offset: Offset(0.0, 0.0),
                            ),
                          ],
                        ),

                      )),
                  opacity: ActorOpacity(opacity: 0)),
              ActorAction(
                  waitBeforeStart: Duration(milliseconds: 5000),
                  opacity: ActorOpacity(
                      opacity: 1, duration: Duration(milliseconds: 3000)))
            ]),
            ActorWidget(
                name: '6',
                actions: [
              ActorAction(
                position: ActorPosition(
                  top: MediaQuery.of(context).size.height - 120,
                  left: -300,
                ),
                character: ActorCharacter(
                    height: 200,
                    width: 300,
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: <Widget>[
                        Text(
                          "The best burgershop in the country!",
                          style: TextStyle(fontSize: 15, color: Colors.black87),
                        ),
                        Text(
                          "Visit us and taste our burgers...",
                          style: TextStyle(fontSize: 15, color: Colors.black87),
                        ),
                      ],
                    )),
              ),
              ActorAction(
                  waitBeforeStart: Duration(milliseconds: 5000),
                  position: ActorPosition(
                    left: 20,

                  ))
            ])*/
      ],
    );
  }

  ActorWidget itemText(
      {Curve curve,
        Duration duration,
        Duration waitBefore,
        String title,
        double top}) {
    return ActorWidget(
        name: title,
        actions: [
          ActorAction(
            waitBeforeStart: waitBefore,
            position: ActorPosition(left: -200, top: top),
            character: ActorCharacter(
                width: 200,
                height: 100,
                child: Text(
                  title,
                  style: TextStyle(color: Colors.white70, fontSize: 25),
                )),
          ),
          ActorAction(
              position: ActorPosition(
                  left: 30,
                  curve: Curves.elasticOut,
                  duration: Duration(milliseconds: 2000)))
        ]);
  }

  ActorWidget optionCircle(
      {int traversePercentage,
        Curve pathCurve,
        Duration pathDuration,
        Duration waitBefore,
        Duration rotateDuration,
        double rotationTurn = 1.0,
        String image}) {
    return ActorWidget(
        name: image,
        actions: [
          ActorAction(
              waitBeforeStart: waitBefore - Duration(milliseconds: 500),
              path: ActorPath(
                  displayPath: false,
                  traversePercentage: traversePercentage,
                  path: 'M -30 100 A 1 1 0 0 1 -30 500 ',
                  curve: pathCurve,
                  duration: pathDuration),
              character: ActorCharacter(
                  width: 100,
                  height: 100,
                  child: Image.asset('assets/images/' + image + '.png')),
              rotation: ActorRotation(
                  rotationTurns: rotationTurn,
//            clockwise: false,
//              horizontalFlipTurns: 1.0,
                  startTween: 0.6,
                  duration: rotateDuration,
                  curve: Curves.elasticOut)),
          ActorAction(
              position: ActorPosition(
                  left: 210,
                  curve: Curves.elasticOut,
                  duration: Duration(milliseconds: 2000)))
        ]);
  }
}


